design-proposals/preferences.md

```diff
@@ -0,0 +1,584 @@
+
+Additionally the preferences would be tightly coupled with the sizing of resources within the flavor. As a result third parties such as application or distribution vendors would not be able to provided generic preferences independently.
+
+In OpenStack for example this issue is mostly avoided through the use of [image properties](https://docs.openstack.org/glance/latest/admin/useful-image-properties.html). This is where the boot image of an instance (`VMI`) is decorated with properties in order to override or augment preferences defined by [flavors](https://docs.openstack.org/nova/latest/user/flavors.html).
+
+As this isn't possible within `KubeVirt` another mechanism is required to separate the quantitative compute resources defined by flavors and qualitative preferences we also want to provide.
+
+## User Stories
+
+* As an operator or user I would like to provide flavors that define the compute resources available to a VM.
+
+* As an operator or user I would like to provide preferences that relate to the performance characteristics of a VM within these flavors.
+
+* As an operator, user or third party (application or distribution vendor) I would like to independently provide preferences that relate to the performance characteristics of a VM outside of a flavor.
+
+## Goals
+
+* Provide a way for both operators, users and third parties to express performance characteristics of a VM while avoiding the need to duplicate flavors by the number of resource sizes being provided within an environment.
+
+## Non-Goals
+
+The following items have been discussed alongside the introduction of Flavors but are not within the current scope of this design proposal:
+
+* Introducing an `optional` policy for some of the preferences as done with OpenStack [Flavor extra-specs](https://docs.openstack.org/nova/latest//configuration/extra-specs.html#hw:cpu_thread_policy). It isn't clear how we would handle scheduling failures and retries here within KubeVirt/k8s.
+
+* Versioning of [`VirtualMachineFlavor`](http://kubevirt.io/api-reference/main/definitions.html#_v1alpha1_virtualmachineflavor),  [`VirtualMachineClusterFlavor`](http://kubevirt.io/api-reference/main/definitions.html#_v1alpha1_virtualmachineclusterflavor), `VirtualMachinePreferences` and `VirtualMachineClusterPreferences`.
+
+* Extracting [`Resources`](http://kubevirt.io/api-reference/main/definitions.html#_k8s_io_api_core_v1_resourcerequirements) from the [`DomainSpec`](http://kubevirt.io/api-reference/main/definitions.html#_v1_domainspec).
+
+* Adding more supported [`Extended Resources`](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#extended-resources) such as [`GPUs`](http://kubevirt.io/api-reference/main/definitions.html#_v1_gpu) and [`HostDevices`](http://kubevirt.io/api-reference/main/definitions.html#_v1_hostdevice) etc.
```

Comment:
It would be nice, if the proposal could show an idea, how this could be modeled, because I guess this is tightly coupled with the way the VM families/types are modeled.

Reply:
I've dropped L41 after a discussion last week about extended resources, I wasn't aware that GPUs etc can be expressed as resource requests.